datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java

```diff
@@ -736,23 +736,36 @@ private void removeTaskNode(String instance, String name) {
    * act on the differences — that is: add new assignments, remove old assignments.
    * For ones that didn't change, do nothing.
    *
+   * Can receive duplicate task assignment. In this case select one of the tasks as
```

Comment:
If I understand this correctly,

In the `updateAllAssignments`, the leader coordinator adds tasks initially ([code link here](https://github.com/linkedin/brooklin/blob/87c7bec012a776cbb4a5be7d40903682be2d763e/datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java#L758)) to the other instance based on the newly generated assignment. But if that leader instance crashes before removing that task from the previously assigned instance, there could be two different instances handling that same task.

And if that is the issue, I would prefer that we have a fix for that in the `performAssignment` method (in all implementations of this method) since I feel that although deduping it in the ZK code path would fix the issue for this case, but I think this should be handled by the `performAssignment` method. In the future, there could be other code paths that could use the `performAssignment` function, and they'll return a bad assignment (with the same tasks in two instances.)

And I think @vmaheshw has already fixed it for the StickyMulticastStrategy (https://github.com/linkedin/brooklin/pull/749) so maybe other strategies need a similar fix, or it's a different issue?

Reply:
I believe it is a bit more nuanced than this. If a leader crash happens while saving an assignment, a new leader could come in and reuse a unique task name and create a whole new DataStreamTask with the same name. Thus the case that the StickyMulticastStrategy solves wouldn't matter here because doing an equality check between the two DataStreamTasks would not pass but a crash would still happen later when we expect a set of unique task names

Reply:
Reusing a unique task name for creating a newer task should be super rare(atleast in all strategies except LoadBalanced), which is not as probable as the issue described in the previous comment I feel, where the same task could land upon multiple hosts.

However, even if that happens, I would prefer the strategy handles cleaning it up instead of cleaning it in the ZK code path.

Reply:
Updated the assignment of each strategy